File: server/apps/nexotype/schemas/user/user_profile_schemas.py
from pydantic import BaseModel, ConfigDict
from datetime import datetime

# ==========================================
# Request Schemas
# ==========================================

class UserProfileCreate(BaseModel):
    """Schema for creating a new user profile"""
    user_id: int
    subject_id: int


class UserProfileUpdate(BaseModel):
    """Schema for updating a user profile"""
    user_id: int = None
    subject_id: int = None


# ==========================================
# Response Schemas
# ==========================================

class UserProfileDetail(BaseModel):
    """Schema for user profile details"""
    id: int
    user_id: int
    subject_id: int
    created_at: datetime
    updated_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


class UserProfileListResponse(BaseModel):
    """Response schema for listing user profiles"""
    success: bool
    data: list[UserProfileDetail] | None = None
    count: int = 0
    error: str | None = None


class UserProfileResponse(BaseModel):
    """Response schema for single user profile operations"""
    success: bool
    data: UserProfileDetail | None = None
    error: str | None = None


class MessageResponse(BaseModel):
    """Simple message response"""
    success: bool
    message: str | None = None
    error: str | None = None

File: client/src/modules/nexotype/schemas/user/user-profile.schemas.ts
import { z } from 'zod';

// ==========================================
// Full Schema (GET responses)
// ==========================================

export const UserProfileSchema = z.object({
  id: z.number(),
  user_id: z.number().int(),
  subject_id: z.number().int(),
  created_at: z.string(),
  updated_at: z.string().nullable(),
});

// ==========================================
// Input Schemas (POST/PUT)
// ==========================================

export const CreateUserProfileSchema = z.object({
  user_id: z.number().int(),
  subject_id: z.number().int(),
});

export const UpdateUserProfileSchema = z.object({
  user_id: z.number().int().optional(),
  subject_id: z.number().int().optional(),
});

// ==========================================
// Types
// ==========================================

export type UserProfile = z.infer<typeof UserProfileSchema>;
export type CreateUserProfile = z.infer<typeof CreateUserProfileSchema>;
export type UpdateUserProfile = z.infer<typeof UpdateUserProfileSchema>;

// ==========================================
// Response Types
// ==========================================

export type UserProfileResponse = {
  success: boolean;
  data?: UserProfile;
  error?: string;
};

export type UserProfilesResponse = {
  success: boolean;
  data?: UserProfile[];
  count?: number;
  error?: string;
};
